docs/language/learn-ql/ql-etudes/river-crossing.rst

```diff
@@ -0,0 +1,243 @@
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing.ql
+      :lines: 15-22
+
+Second, any item can be on one of two shores. Let's call these the "left shore" and the "right shore".
+Define a class ``Shore`` containing ``"Left"`` and ``"Right"``.
+
+It would be helpful to express "the other shore". You can do this by defining a member predicate
+``flip`` in the class ``Shore`` such that ``"Left".flip()`` returns ``"Right"`` and vice versa.
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing.ql
+      :lines: 25-38
+
+We also want a way to keep track of where the man, the goat, the cabbage, and the wolf are—call this combined
+information the "state". Define a class ``State`` that encodes this information.
+For example, if the man is on the left shore, the goat on the right shore, and the cabbage and wolf on the left
+shore, the state should be ``Left, Right, Left, Left``.
+
+You may find it helpful to introduce some variables that refer to the shore on which the man and the cargo items are. These
+temporary variables in the body of a class are called `fields <https://help.semmle.com/QL/ql-handbook/types.html#fields>`__.
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing-1.ql
+      :lines: 32-42,83
+
+We are interested in two particular states, namely the initial state and the goal state.
+Assuming that all items start on the left shore and end up on the right shore, define
+``InitialState`` and ``GoalState`` as subclasses of ``State``.
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing-1.ql
+      :lines: 85-93
+
+.. pull-quote::
+
+   Note
+
+   To avoid typing out the lengthy string concatenations, you could introduce a helper predicate
+   ``renderState`` that renders the state in the required form.
+
+Using the above note, the QL code so far looks like this:
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing.ql
+      :lines: 14-55,100-110
+
+Model the action of "ferrying"
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The basic act of ferrying moves the man and a cargo item to the other shore,
+resulting in a new state.
+
+Write a member predicate (of ``State``) called ``ferry``, that specifies what happens to the state
+after ferrying a particular cargo. (Hint: Use the predicate ``flip``.)
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing.ql
+      :lines: 57-66
+
+Of course, not all ferrying actions are possible. Add some extra conditions to describe when a ferrying
+action is "safe", that is, it doesn't lead to a state where the goat or the cabbage get eaten.
+For example, follow these steps:
+
+   #. Define a predicate ``eats`` that encodes the conditions for when something gets eaten.
```

Comment:
This formulation does not make me think of a predicate with two arguments for predator and prey. Maybe something along the lines of "...for when a predator would be able to eat an unguarded prey" would provide a stronger hint. (Even though predator and prey also seem slightly wrong for the goat-cabbage situation.)